applications/crossbar/src/modules/cb_alerts.erl

```diff
@@ -230,19 +230,26 @@ maybe_check_port_requests(Context) ->
     AccountId = cb_context:account_id(Context),
 
     case {is_reseller_handling_port_requests(ResellerId)
-         ,ResellerId =:= AccountId
+         ,kz_services_reseller:is_reseller(AccountId)
```

Comment:
I did rewrite this `maybe_check_port_requests` below, so it it won't do descendant for master account and it kind of more readable?? What do you think @harenson ?

This is diff, you can apply it by saving to a file and do `git apply path_to_diff_file`.

```diff
diff --git a/applications/crossbar/src/modules/cb_alerts.erl b/applications/crossbar/src/modules/cb_alerts.erl
index 21cdd4ab3f..5303f03a08 100644
--- a/applications/crossbar/src/modules/cb_alerts.erl
+++ b/applications/crossbar/src/modules/cb_alerts.erl
@@ -226,48 +226,50 @@ set_success_resp_status(Context) ->
 %%------------------------------------------------------------------------------
 -spec maybe_check_port_requests(cb_context:context()) -> cb_context:context().
 maybe_check_port_requests(Context) ->
+    MasterId = cb_context:master_account_id(Context),
     ResellerId = cb_context:reseller_id(Context),
     AccountId = cb_context:account_id(Context),
-
-    case {is_reseller_handling_port_requests(ResellerId)
-         ,kz_services_reseller:is_reseller(AccountId)
-          andalso is_reseller_handling_port_requests(AccountId)
-         }
-    of
-        {'true', _} ->
-            %% Reseller is handling port requests so not bother the user with port request alerts.
-            Context;
-        {'false', 'true'} ->
-            %% Reseller is NOT handling port requests for sub-accounts and current request
-            %% was made by a reseller and it is handling port requests for its sub-accounts
-            %% so get the alerts (if any) for each sub-account.
+    maybe_check_port_requests(Context, AccountId, ResellerId, MasterId).
+
+-spec maybe_check_port_requests(cb_context:context(), kz_term:api_ne_binary(), kz_term:api_ne_binary(), kz_term:api_ne_binary()) -> cb_context:context().
+maybe_check_port_requests(Context, MasterId, _, MasterId) ->
+    %% showing master account's ports
+    do_check_port_requests(fetch_account_active_ports(MasterId), Context);
+maybe_check_port_requests(Context, ResellerId, ResellerId, _) ->
+    case should_hide_port(ResellerId) of
+        'true' ->
+            %% Reseller is handling port requests for sub-accounts, show ports.
             do_check_port_requests(knm_port_request:descendant_active_ports(ResellerId), Context);
-        {'false', _} ->
-            %% Reseller is not handling port requests and either the current account is
-            %% not a reseller or it is not handling port requests for its sub-accounts.
-            Ports = maybe_merge_active_and_unconfirmed_ports(
-                      knm_port_request:account_active_ports(AccountId)
-                     ,knm_port_request:account_ports_by_state(AccountId, ?PORT_UNCONFIRMED)
-                     ),
-            do_check_port_requests(Ports, Context)
+        'false' ->
+            Context
+    end;
+maybe_check_port_requests(Context, AccountId, ResellerId, _) ->
+    case should_hide_port(ResellerId) of
+        'true' -> Context;
+        'false' ->
+            do_check_port_requests(fetch_account_active_ports(AccountId), Context)
     end.
 
--spec is_reseller_handling_port_requests(kz_term:api_ne_binary() | {'ok', kz_json:object()} | {'error', any()}) -> boolean().
-is_reseller_handling_port_requests('undefined') ->
+-spec should_hide_port(kz_term:api_ne_binary() | {'ok', kz_json:object()} | {'error', any()}) -> boolean().
+should_hide_port('undefined') ->
     'false';
-is_reseller_handling_port_requests(<<_/binary>> = ResellerId) ->
-    is_reseller_handling_port_requests(kzd_whitelabel:fetch(ResellerId));
-is_reseller_handling_port_requests({'ok', Whitelabel}) ->
-    kz_json:is_true(<<"hide_port">>, Whitelabel);
-is_reseller_handling_port_requests({'error', 'not_found'}) ->
+should_hide_port(<<ResellerId/binary>>) ->
+    should_hide_port(kzd_whitelabel:fetch(ResellerId));
+should_hide_port({'ok', Whitelabel}) ->
+    kzd_whitelabel:hide_port(Whitelabel);
+should_hide_port({'error', 'not_found'}) ->
     'false'.
 
--spec maybe_merge_active_and_unconfirmed_ports({'ok', kz_json:objects()} | {'error', 'not_found'}
-                                              ,{'ok', kz_json:objects()} | {'error', 'not_found'}
-                                              ) -> kz_json:objects().
-maybe_merge_active_and_unconfirmed_ports({'ok', Active}, {'ok', Unconfirmed}) ->
-    lists:flatten([Active, Unconfirmed]);
-maybe_merge_active_and_unconfirmed_ports(_, _) ->
+-spec fetch_account_active_ports(kz_term:api_ne_binary()) -> kz_json:objects().
+fetch_account_active_ports('undefined') -> [];
+fetch_account_active_ports(AccountId) ->
+    get_active_ports(knm_port_request:account_active_ports(AccountId))
+        ++ get_active_ports(knm_port_request:account_ports_by_state(AccountId, ?PORT_UNCONFIRMED)).
+
+-spec get_active_ports({'ok', kz_json:objects()} | {'error', 'not_found'}) -> kz_json:objects().
+get_active_ports({'ok', Active}) ->
+    Active;
+get_active_ports(_) ->
     [].
 
 -spec do_check_port_requests(kz_term:api_ne_binary() | kz_json:objects(), cb_context:context()) -> cb_context:context().
```

Reply:
Ok, so this slightly change propose for the above diff/patch:

So the decision for showing the alert is as follow:

* If the authenticated account is **master** account:
  * Show any port's alerts for the masqueraded account (the account id from URL or in other word `cb_context:account_id`)
* If the authenticated account is _re-seller_ **and** _port authority_ :
  * Show all alerts from the reseller account. Because the reseller is the port authority of its sub-accounts, so it was him that rejected sub-accounts ports or made action required  comment on sub-account's ports, and it should not being alerted for his own action.
* If the authenticated account is _re-seller_:
  * show all alerts from reseller account and its sub-accounts, no matter what account id the request is masquerading
* for regular authenticated account, so alerts for the account id in URL (`cb_context:account_id`)

This is a rough implementation of above bullets, please check and test the code for its logic and/or for any missed-typed or other issues.

```diff
diff --git a/applications/crossbar/src/modules/cb_alerts.erl b/applications/crossbar/src/modules/cb_alerts.erl
index 21cdd4ab3f..64c188e43f 100644
--- a/applications/crossbar/src/modules/cb_alerts.erl
+++ b/applications/crossbar/src/modules/cb_alerts.erl
@@ -226,48 +226,65 @@ set_success_resp_status(Context) ->
 %%------------------------------------------------------------------------------
 -spec maybe_check_port_requests(cb_context:context()) -> cb_context:context().
 maybe_check_port_requests(Context) ->
-    ResellerId = cb_context:reseller_id(Context),
     AccountId = cb_context:account_id(Context),
+    MasterId = cb_context:master_account_id(Context),
+    AuthId = cb_context:auth_account_id(Context),
+    case AuthId =:= MasterId of
+        'true' ->
+            %% Authenticated account is master
+            %% show only masqueraded account's port alerts
+            do_check_port_requests(fetch_account_active_ports(AccountId), Context);
+        'false' ->
+            IsReseller = kz_services_reseller:is_reseller(AccountId),
+            IsPortAuthority = cb_context:fetch(Context, 'is_port_authority'),
+            maybe_check_port_requests(Context, IsReseller, IsPortAuthority)
+    end.-    case {is_reseller_handling_port_requests(ResellerId)
-         ,kz_services_reseller:is_reseller(AccountId)
-          andalso is_reseller_handling_port_requests(AccountId)
-         }
-    of
-        {'true', _} ->
-            %% Reseller is handling port requests so not bother the user with port request alerts.
-            Context;
-        {'false', 'true'} ->
-            %% Reseller is NOT handling port requests for sub-accounts and current request
-            %% was made by a reseller and it is handling port requests for its sub-accounts
-            %% so get the alerts (if any) for each sub-account.
-            do_check_port_requests(knm_port_request:descendant_active_ports(ResellerId), Context);
-        {'false', _} ->
-            %% Reseller is not handling port requests and either the current account is
-            %% not a reseller or it is not handling port requests for its sub-accounts.
-            Ports = maybe_merge_active_and_unconfirmed_ports(
-                      knm_port_request:account_active_ports(AccountId)
-                     ,knm_port_request:account_ports_by_state(AccountId, ?PORT_UNCONFIRMED)
-                     ),
-            do_check_port_requests(Ports, Context)
+-spec maybe_check_port_requests(cb_context:context(), boolean(), boolean()) -> cb_context:context().
+maybe_check_port_requests(Context, 'true', 'true') ->
+    %% Authenticated account is reseller AND port authority
+    %% so show only masqueraded account's port alerts
+    %% (because reseller is port authority, this is whose rejected ports or made action required
+    %% comments on the ports, so it shouldn't get alerted for its own doing!)
+    do_check_port_requests(fetch_account_active_ports(cb_context:account_id(Context)), Context);
+maybe_check_port_requests(Context, 'true', 'false') ->
+    %% Authenticated account is reseller lets always show alerts
+    %% from its own account and all sub-accounts (no matter what port app is hidden or not)
+    ResellerId = cb_context:reseller_id(Context),
+    Ports = fetch_account_active_ports(ResellerId)
+                ++ get_active_ports(knm_port_request:descendant_active_ports(ResellerId)),
+    do_check_port_requests(Ports, Context);
+maybe_check_port_requests(Context, 'false', _) ->
+    %% Authenticated account is neither master or reseller
+    %% show alerts if its reseller is not hiding port app
+    AccountId = cb_context:account_id(Context),
+    ResellerId = cb_context:reseller_id(Context),
+    case should_hide_port(ResellerId) of
+        'true' -> Context;
+        'false' ->
+            do_check_port_requests(fetch_account_active_ports(AccountId), Context)
     end.--spec is_reseller_handling_port_requests(kz_term:api_ne_binary() | {'ok', kz_json:object()} | {'error', any()}) -> boolean().
-is_reseller_handling_port_requests('undefined') ->
+-spec should_hide_port(kz_term:api_ne_binary() | {'ok', kz_json:object()} | {'error', any()}) -> boolean().
+should_hide_port('undefined') ->
     'false';
-is_reseller_handling_port_requests(<<_/binary>> = ResellerId) ->
-    is_reseller_handling_port_requests(kzd_whitelabel:fetch(ResellerId));
-is_reseller_handling_port_requests({'ok', Whitelabel}) ->
-    kz_json:is_true(<<"hide_port">>, Whitelabel);
-is_reseller_handling_port_requests({'error', 'not_found'}) ->
+should_hide_port(<<ResellerId/binary>>) ->
+    should_hide_port(kzd_whitelabel:fetch(ResellerId));
+should_hide_port({'ok', Whitelabel}) ->
+    kzd_whitelabel:hide_port(Whitelabel);
+should_hide_port({'error', 'not_found'}) ->
     'false'.--spec maybe_merge_active_and_unconfirmed_ports({'ok', kz_json:objects()} | {'error', 'not_found'}
-                                              ,{'ok', kz_json:objects()} | {'error', 'not_found'}
-                                              ) -> kz_json:objects().
-maybe_merge_active_and_unconfirmed_ports({'ok', Active}, {'ok', Unconfirmed}) ->
-    lists:flatten([Active, Unconfirmed]);
-maybe_merge_active_and_unconfirmed_ports(_, _) ->
+-spec fetch_account_active_ports(kz_term:api_ne_binary()) -> kz_json:objects().
+fetch_account_active_ports('undefined') -> [];
+fetch_account_active_ports(AccountId) ->
+    get_active_ports(knm_port_request:account_active_ports(AccountId))
+        ++ get_active_ports(knm_port_request:account_ports_by_state(AccountId, ?PORT_UNCONFIRMED)).
+
+-spec get_active_ports({'ok', kz_json:objects()} | {'error', 'not_found'}) -> kz_json:objects().
+get_active_ports({'ok', Active}) ->
+    Active;
+get_active_ports(_) ->
     []. -spec do_check_port_requests(kz_term:api_ne_binary() | kz_json:objects(), cb_context:context()) -> cb_context:context().
```